193 lesen und schreiben von ausdrucken

dragonfly28 opened the pull request:
# Beschreibung:

Speichern und Lesen von Ausdrucken.

Der Legacy-Code hatte einige Unschönheiten, die aber überwiegend so übernommen wurden, z.B.
- keine Authorities im Repository
- inkonsistente HTTP-Return-Codes im Controller
- keine Optionals im Repo
Bei Verbesserungswünschen bitte entsprechende Anmerkungen machen!


<!-- Backend -->
### Backend ###
- [x] [Codingconventions](https://it-at-m.github.io/Wahllokalsystem/technik/coding_conventions/) beachtet
- [x] Doku aktualisiert
- [x] Swagger-API vollständig
- [x] Unit-Tests gepflegt
- [x] Integrationstests gepflegt
- [x] Beispiel-Requests gepflegt

<!-- Dokumentation -->
### Dokumentation
- [x] Links geprüft

Closes #193


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

Based on the comprehensive changes, here are the release notes:

- **Documentation**
	- Added documentation for Ergebnismeldung-Service, including details about handling printed documents
	- New sections describing reading and writing of printed documents for election results

- **New Features**
	- Introduced functionality to save and retrieve printed documents for specific electoral districts
	- Added support for two types of documents: Niederschrift (V1) and Schnellmeldung (V3)
	- Implemented document management with creation date tracking

- **Security**
	- Added new permissions for retrieving and posting documents
	- Enhanced access control for document-related operations

- **Database**
	- Created new database table `Ausdruck` to store printed documents
	- Implemented composite key structure for document identification

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dragonfly28 committed:
domain classes and repo for 'Ausdruck'

dragonfly28 committed:
repository test and models

dragonfly28 committed:
DTOs, controller and service

dragonfly28 committed:
- LocalDateTimeComparators renamed
- all precision comparators for date and time moved to this class;

dragonfly28 committed:
validator and mapper tests

dragonfly28 committed:
more tests

dragonfly28 committed:
service test

dragonfly28 committed:
service security test and mvn spotless

dragonfly28 committed:
some refactoring and bugs

dragonfly28 committed:
controller test added

dragonfly28 committed:
controller integration test added

dragonfly28 committed:
test added

dragonfly28 committed:
swagger doc added

dragonfly28 committed:
permissions added to auth-service

dragonfly28 committed:
HTTP test requests added

dragonfly28 committed:
Merge branch 'dev' into 193-lesen-und-schreiben-von-ausdrucken

dragonfly28 committed:
version changed after merging dev

dragonfly28 committed:
adapted to new auth-service

dragonfly28 committed:
doc structure added

dragonfly28 committed:
doc added

dragonfly28 committed:
naming conventions applied

dragonfly28 committed:
refactoring for better naming

dragonfly28 committed:
close MockedStatic after usage

dragonfly28 committed:
checkstyle violation corrected

dragonfly28 commented:
@coderabbitai review

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/WahlUndBezirkIDUndMeldungsartValidator.java:
```diff
@@ -0,0 +1,22 @@
+@RequiredArgsConstructor
+public class WahlUndBezirkIDUndMeldungsartValidator {
+
+    public void validWahlUndBezirkIDUndMeldungsartOrThrow(WahlUndBezirkIDUndMeldungsart wahlUndBezirkIDUndMeldungsart,
+            final FachlicheWlsException exceptionOnInvalid)
+            throws FachlicheWlsException {
+        if (StringUtils.isBlank(wahlUndBezirkIDUndMeldungsart.getWahlID()) ||
+                StringUtils.isBlank(wahlUndBezirkIDUndMeldungsart.getWahlbezirkID()) ||
+                wahlUndBezirkIDUndMeldungsart.getMeldungsart() == null) {
+            throw exceptionOnInvalid;
+        }
+    }
```
Exception wird von aussen mit Fehlercode mitgegeben wie im Legacy-System

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckService.java:
```diff
@@ -0,0 +1,65 @@
+        if (result == null) {
+            log.info("Printout not found for: {}", id);
+        }
+        return ausdruckModelMapper.toModel(result);
```
keine Änderung - Legacy-Code loggt nur, ModelMapper gibt null zurück

dragonfly28 commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/utils/Authorities.java:
```diff
@@ -17,6 +17,8 @@ public class Authorities {
     public static final String SERVICE_SET_BEGRUENDUNG = "Ergebnismeldung_BUSINESSACTION_PostBegruendung";
     public static final String SERVICE_GET_STIMMZETTELUMSCHLAEGE = "Ergebnismeldung_BUSINESSACTION_GetStimmzettelumschlaege";
     public static final String SERVICE_SET_STIMMZETTELUMSCHLAEGE = "Ergebnismeldung_BUSINESSACTION_PostStimmzettelumschlaege";
+    public static final String SERVICE_GET_AUSDRUCK = "Ergebnismeldung_BUSINESSACTION_GetAusdruck";
+    public static final String SERVICE_POST_AUSDRUCK = "Ergebnismeldung_BUSINESSACTION_PostAusdruck";
```
Wäre schön, ist aber im Legacy-Code auch nicht vorhanden.

dragonfly28 commented on wls-auth-service/src/main/resources/db/migrations/h2/V6_0__addErgebnismeldungPermissionsAusdruck.sql:
```diff
@@ -0,0 +1,9 @@
+INSERT INTO Permission
+VALUES ('00000000-0000-0006-0000-000000000001', 'Ergebnismeldung_BUSINESSACTION_GetAusdruck');
+INSERT INTO Secauthorities_Secpermissions
+VALUES ('00000000-0000-0000-0001-000000000001', '00000000-0000-0006-0000-000000000001');
+
+INSERT INTO Permission
+VALUES ('00000000-0000-0006-0000-000000000002', 'Ergebnismeldung_BUSINESSACTION_PostAusdruck');
+INSERT INTO Secauthorities_Secpermissions
+VALUES ('00000000-0000-0000-0001-000000000001', '00000000-0000-0006-0000-000000000002');
```
nicht notwendig - macht flyway

dragonfly28 commented on docs/src/services/ergebnismeldung-service/index.md:
```diff
@@ -61,4 +61,17 @@ bzw. ausgelesen werden.
\ No newline at end of file
+übermittelt bzw. ausgelesen werden.
+
+### Lesen und Schreiben von Ausdrucken
+Hier können Niederschrift(V1) und Schnellmeldung(V3) gespeichert und gelesen werden.
+
+####  Speichern eines Ausdrucks mit bestimmter Meldungsart für einen Wahlbezirk einer Wahl
+Speichert einen Ausdruck der entsprechenden Meldungsart für einen Wahlbezirk einer Wahl.
+Es wird automatisch ein aktuelles Erstellungsdatum mit erzeugt und gespeichert.
+
+### Lesen eines Ausdrucks mit bestimmter Meldungsart für einen Wahlbezirk einer Wahl
+Liest einen Ausdruck mit entsprechender Meldungsart für einen Wahlbezirk einer Wahl.
+
+### Lesen aller Ausdrucke für einen Wahlbezirk einer Wahl
+Liest alle Ausdrucke (V1 und V3) für einen Wahlbezirk einer Wahl.
```
Das ist eher fachliche Doku, die technische kann man ja in der swagger-ui ansehen

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/domain/ausdruck/Ausdruck.java:
```diff
@@ -0,0 +1,31 @@
+    @EmbeddedId
+    private WahlUndBezirkIDUndMeldungsart wahlUndBezirkIDUndMeldungsart;
+
+    @Lob
+    private String content;
```
keine Änderung, bleibt wie es ist

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/domain/ausdruck/AusdruckRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface AusdruckRepository extends CrudRepository<Ausdruck, WahlUndBezirkID> {
+
+    Ausdruck findOneByWahlUndBezirkIDUndMeldungsart(WahlUndBezirkIDUndMeldungsart id);
```
keine Änderung - Legacy-Code übernommen

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckService.java:
```diff
@@ -0,0 +1,65 @@
+    public void saveAusdruck(@P("param") @NotNull final AusdruckModel ausdruck) {
+        log.debug("Saving printout {}", ausdruck.wahlUndBezirkIDUndMeldungsart().getMeldungsart());
+
+        if (!validator.validate(ausdruck).isEmpty()) {
+            throw exceptionFactory.createFachlicheWlsException(ExceptionConstants.POST_AUSDRUCK_PARAMETER_UNVOLLSTAENDIG);
+        }
```
Keine Änderung - Legacy-Code

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/domain/ausdruck/AusdruckRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+import org.springframework.data.repository.CrudRepository;
+
+public interface AusdruckRepository extends CrudRepository<Ausdruck, WahlUndBezirkID> {
```
Anstalle von `WahlUndBezirkID` würde ich die Klasse der ID verwenden: `WahlUndBezirkIDUndMeldungsart`. Entsprechend ist kein zusätzliche Methode `findOneByWahlUndBezirkIDUndMeldungsart` erforderlich sondern es kann `findById` verwendet werden.

dragonfly28 replied:
resolved with e2988c29a67e425023802df6ff0d5870accd2357

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/domain/ausdruck/AusdruckRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+    Ausdruck findOneByWahlUndBezirkIDUndMeldungsart(WahlUndBezirkIDUndMeldungsart id);
+
+    List<Ausdruck> findAllByWahlUndBezirkIDUndMeldungsart_WahlIDAndWahlUndBezirkIDUndMeldungsart_WahlbezirkID(String wahlId, String wahlbezirkId);
```
Den Namen finde ich nicht besonders leserlich. Mit @GerhardPx hatte ich bei Ergebnisse darüber diskutiert. Ich finde eine Lösung mit einer `@Query`-Annotation besser:

```java
@Query(...)
List<Ausdruck> findByWahlIdAndWahlbezirkId(String wahlId, String wahlbezirkId);
```

dragonfly28 replied:
resolved with e3a4c7f9d78e95c2699513de9667ea9290cbfc5d

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckController.java:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        val responseHeaders = new HttpHeaders();
+        responseHeaders.add(HttpHeaders.CONTENT_TYPE, "text/html; charset=utf-8");
```
Hierzu bitte ein Issue erstellen. Wenn die Anfrage mit einem anderen Zeichensatz kodiert ist ist die Antwort unsauber. Speichern und Lesen müssen kompatibel sein.

dragonfly28 replied:
Issue erstellt -> https://github.com/it-at-m/Wahllokalsystem/issues/747

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckReadDTO.java:
```diff
@@ -0,0 +1,12 @@
+package de.muenchen.oss.wahllokalsystem.ergebnismeldungservice.rest.ausdruck;
+
+import de.muenchen.oss.wahllokalsystem.ergebnismeldungservice.domain.ausdruck.Meldungsart;
```
Klassen aus dem rest-Package sollten nicht auf andere spezifische Datenmodell verweisen.

dragonfly28 replied:
resolved with 0091a4e78a9ed7c7547f6482feded460ede46397

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckController.java:
```diff
@@ -0,0 +1,129 @@
+    @GetMapping("{wahlID}/{wahlbezirkID}")
+    public ResponseEntity<List<AusdruckReadDTO>> getAllAusdrucke(@PathVariable("wahlID") final String wahlID,
+            @PathVariable("wahlbezirkID") final String wahlbezirkID) {
+        List<AusdruckReadDTO> res = ausdruckReadDTOMapper.fromListOfAusdruckModelToListOfAusdruckReadDTO(ausdruckService.getAllAusdrucke(wahlID, wahlbezirkID));
```
an Stelle des expliziten Mapping-Methode könnte man mit `stream.map.toList` arbeiten. Damit könnte man sich einen längeren Methodennamen sparen.

dragonfly28 replied:
resolved with 14e2e14a46614412b2ea4e884c55264b18f16fb2

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckController.java:
```diff
@@ -0,0 +1,129 @@
+            @PathVariable("wahlbezirkID") final String wahlbezirkID, @PathVariable("meldungsart") final Meldungsart meldungsart,
+            @RequestBody final AusdruckWriteDTO ausdruck) {
+        AusdruckModel ausdruckModel = ausdruckWriteDTOMapper.toModel(ausdruck, new WahlUndBezirkIDUndMeldungsart(wahlbezirkID, wahlID, meldungsart),
+                Instant.now());
```
Die Entscheidung `erstelltAm` `Instant.now` ist sollte im Service erfolgen da es fachlich ist.

dragonfly28 replied:
resolved with 68b3e359fb1de4c86c618c37a48fb72216db90b1

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckService.java:
```diff
@@ -0,0 +1,65 @@
+
+    @PreAuthorize("hasAuthority('Ergebnismeldung_BUSINESSACTION_GetAusdruck')")
+    public List<AusdruckModel> getAllAusdrucke(@NotBlank final String wahlID, @NotBlank final String wahlbezirkID) {
+        return ausdruckModelMapper.toModelList(
```
gleicher Vorschlag wie beim Controller hier mit stream anstatt einer List-Mapper-Methode zu arbeiten.

dragonfly28 replied:
resolved with 14e2e14a46614412b2ea4e884c55264b18f16fb2

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckService.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @PreAuthorize("hasAuthority('Ergebnismeldung_BUSINESSACTION_GetAusdruck')")
+    public AusdruckModel getAusdruck(WahlUndBezirkIDUndMeldungsart id) {
```
Kannst du das auch auf `Optional<AusdruckModel>` ändern?

dragonfly28 replied:
resolved with e2988c29a67e425023802df6ff0d5870accd2357

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/WahlUndBezirkIDUndMeldungsartValidator.java:
```diff
@@ -0,0 +1,22 @@
+    public void validWahlUndBezirkIDUndMeldungsartOrThrow(WahlUndBezirkIDUndMeldungsart wahlUndBezirkIDUndMeldungsart,
+            final FachlicheWlsException exceptionOnInvalid)
+            throws FachlicheWlsException {
+        if (StringUtils.isBlank(wahlUndBezirkIDUndMeldungsart.getWahlID()) ||
```
Hier bitte noch einen Null-Check für `wahlUndBezirkIDUndMeldungsart` ergänzen.

dragonfly28 replied:
resolved with 87af6615329446867f9689b10b57c1ebcbaee183

MrSebastian commented on wls-ergebnismeldung-service/src/main/resources/db/migrations/h2/V6_0__createTableAusdruck.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE Ausdruck
+(
+    wahlID varchar(255) NOT NULL,
+    wahlbezirkID varchar(255) NOT NULL,
+    meldungsart varchar(255) NOT NULL,
+    content CLOB,
+    erstellt_am TIMESTAMP NOT NULL,
+    PRIMARY KEY (wahlID, wahlbezirkID, meldungsart)
```
```suggestion
    wahlID VARCHAR(255) NOT NULL,
    wahlbezirkID VARCHAR(255) NOT NULL,
    meldungsart VARCHAR(255) NOT NULL,
    content CLOB,
    erstellt_am TIMESTAMP NOT NULL,
    PRIMARY KEY (wahlID, wahlbezirkID, meldungsart)
```

MrSebastian commented on wls-ergebnismeldung-service/src/main/resources/db/migrations/oracle/V6_0__createTableAusdruck.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE Ausdruck
+(
+    wahlID varchar(255) NOT NULL,
+    wahlbezirkID varchar(255) NOT NULL,
+    meldungsart varchar(255) NOT NULL,
```
```suggestion
    wahlID VARCHAR(255) NOT NULL,
    wahlbezirkID VARCHAR(255) NOT NULL,
    meldungsart VARCHAR(255) NOT NULL,
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    @Nested
+    class GetAusdruck {
```
Bitte noch einen exemplarischen Test ergänzen der zeigt dass ein 400er kommt wenn die Infrage invalide ist.

dragonfly28 replied:
resolved with e2988c29a67e425023802df6ff0d5870accd2357

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+        }
+
+        @Test
+        void should_returnInternalServerError_when_requestPathParameterWahlIDIsInvalid() throws Exception {
```
warum kommt bei einer invaliden wahlID, wahlbezirkID oder meldungsart ein 500er und kein 400er?

dragonfly28 replied:
Lag am fehlenden @Valid im AusdruckModel
resolved with e2988c29a67e425023802df6ff0d5870accd2357

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    @Nested
+    class GetAllAusdrucke {
+        @Test
```
```suggestion
    class GetAllAusdrucke {

        @Test
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    @Nested
+    class PostAusdruck {
```
Bitte einen Test ergänzen der zeigt dass erfolgreich überschrieben wird.

dragonfly28 replied:
resolved with fea66c9ee10001513c93906b11e601dfdabc82b2

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+        void should_persistData_when_dataIsSent() throws Exception {
+            var clock = Clock.fixed(Instant.now(), ZoneOffset.UTC);
+            var mockedInstant = Instant.now(clock);
+            MockedStatic<Instant> mockedStatic = mockStatic(Instant.class, Mockito.CALLS_REAL_METHODS);
```
Kannst du bitte ein Issue erstellen um es in anderen Service nachzuziehen. Mindestens im Monitoring-Service können wir es anwenden da bei ihm `now` an den Client übermittelt wird.

dragonfly28 replied:
Issue erstellt -> https://github.com/it-at-m/Wahllokalsystem/issues/748

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+    @Nested
+    class GetAllAusdrucke {
+        @Test
+        void should_returnData_when_dataIsPresentInRepository() throws Exception {
+
+            val wahlID = "wahlID";
```
```suggestion
        void should_returnData_when_dataIsPresentInRepository() throws Exception {
            val wahlID = "wahlID";
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+            Mockito.verify(ausdruckService).saveAusdruck((mockedAusdruckModel));
+
+            mockedStatic.close();
+        }
+
+    }
```
```suggestion
        }
    }
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+
+            unitUnderTest.postAusdruck(wahlID, wahlbezirkID, meldungsArt, ausdruckWriteDTO);
+
+            Mockito.verify(ausdruckService).saveAusdruck((mockedAusdruckModel));
```
```suggestion
            Mockito.verify(ausdruckService).saveAusdruck(mockedAusdruckModel);
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+            val mockedMappedServiceDTO = List.of(new AusdruckReadDTO(wahlbezirkID, wahlID, Meldungsart.V1, content, erstelltAm),
+                    new AusdruckReadDTO(wahlbezirkID, wahlID, Meldungsart.V3, content, erstelltAm));
+
+            Mockito.when(ausdruckService.getAllAusdrucke(wahlID, wahlbezirkID)).thenReturn(mockedServiceModel);
+            Mockito.when(ausdruckReadDTOMapper.fromListOfAusdruckModelToListOfAusdruckReadDTO(mockedServiceModel)).thenReturn(mockedMappedServiceDTO);
```
es sollte in der Klasse einheitlich sein. Das wir hier mit einem static-Import arbeiten stört mich wenig

```suggestion
            when(ausdruckService.getAllAusdrucke(wahlID, wahlbezirkID)).thenReturn(mockedServiceModel);
            when(ausdruckReadDTOMapper.fromListOfAusdruckModelToListOfAusdruckReadDTO(mockedServiceModel)).thenReturn(mockedMappedServiceDTO);
```

dragonfly28 replied:
Vorschlag abgelehnt - statt dessen static Import entfernt
siehe fe3a166784dc690ee3da5b6e60c552113adbef52

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+            Assertions.assertThat(result.getStatusCode()).isEqualTo(HttpStatus.OK);
+            Assertions.assertThat(result.getBody()).isSameAs(mockedMappedServiceDTO);
+        }
+    }
+
+}
```
```suggestion
    }
}
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckReadDTOMapperTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Nested
+    class ToDTOList {
```
```suggestion
    class FromListOfAusdruckModelToListOfAusdruckReadDTO {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckReadDTOMapperTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Nested
+    class ToDTOList {
+        @Test
```
```suggestion
    class ToDTOList {

        @Test
```

dragonfly28 replied:
siehe 62e8c57bdda7355220f6b1958906774a464cc1cf

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Nested
+    class ToModelist {
```
```suggestion
    class ToModelList {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    private List<AusdruckModel> createListOfAusdruckModels(Instant erstelltAm) {
```
```suggestion
    private List<AusdruckModel> createListOfAusdruckModels(final Instant erstelltAm) {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java:
```diff
@@ -0,0 +1,130 @@
+
+    }
+
+    private List<Ausdruck> createListOfAusdruckEntities(Instant erstelltAm) {
```
```suggestion
    private List<Ausdruck> createListOfAusdruckEntities(final Instant erstelltAm) {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckServiceSecurityTest.java:
```diff
@@ -0,0 +1,129 @@
+
+        @Test
+        @WithMockUser
+        void should_throwAccessDeniedException_when_serviceGetAusdruckAuthorityIsMissing() {
+
+            val id = new WahlUndBezirkIDUndMeldungsart("wahlbezirkID", "wahlID", Meldungsart.V1);
```
```suggestion
        void should_throwAccessDeniedException_when_serviceGetAusdruckAuthorityIsMissing() {
            val id = new WahlUndBezirkIDUndMeldungsart("wahlbezirkID", "wahlID", Meldungsart.V1);
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/WahlUndBezirkIDUndMeldungsartValidatorTest.java:
```diff
@@ -0,0 +1,49 @@
+    WahlUndBezirkIDUndMeldungsartValidator unitUnderTest;
+
+    @Nested
+    class validWahlUndBezirkIDUndMeldungsartOrThrow {
```
Bitte einen positiven Testfall ergänzen der zeigt dass keine Exception kommt wenn das Argument korrekt ist

dragonfly28 replied:
resolved with 4bc344d298a2042208b6ef86917d088fc48e5938

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/utils/TimePrecisionComparators.java:
```diff
@@ -0,0 +1,13 @@
+
+    public static final Comparator<LocalDateTime> LOCAL_DATE_TIME_PRECISION_MILLISECONDS = Comparator
+            .comparing(dateTime -> dateTime.truncatedTo(ChronoUnit.MILLIS));
+    public static final Comparator<Instant> INSTANT_PRECISION_MILLISECONDS = Comparator.comparing(i -> i.truncatedTo(ChronoUnit.MILLIS));
```
da `i` auch als Zählvariable üblich ist bitte einen anderen Namen wählen. (mein Vorschlag kommt eventuell nicht direkt durch spotless durch

```suggestion
    public static final Comparator<Instant> INSTANT_PRECISION_MILLISECONDS = Comparator.comparing(instant -> instant.truncatedTo(ChronoUnit.MILLIS));
```

MrSebastian reviewed (changes requested):
du hast an diversern Stelle nicht mit `final` bzw. `val` gearbeitet. Bitte da noch mal durch deine Änderungen gehen.

Bitte noch die SecurityConfigurationTests aktualisieren

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/main/resources/db/migrations/h2/V6_0__createTableAusdruck.sql

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/main/resources/db/migrations/oracle/V6_0__createTableAusdruck.sql

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
static import removed

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckReadDTOMapperTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
leerzeile eingefügt

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModelMapperTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckServiceSecurityTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/utils/TimePrecisionComparators.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Merge remote-tracking branch 'origin/193-lesen-und-schreiben-von-ausdrucken' into 193-lesen-und-schreiben-von-ausdrucken

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,136 @@
+
+            Mockito.verify(ausdruckService).saveAusdruck(mockedAusdruckModel);
+
+            mockedStatic.close();
```
bitte das Closing in ein finally oder die Resource in try-catch-resource kapseln:
- https://www.baeldung.com/mockito-mock-static-methods#mocking-a-no-argument-static-method
- https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

dragonfly28 replied:
resolved with db9388fd6cc145ba7e39d6c9663af11bf671c068

dragonfly28 committed:
review annotations:
- WahlUndBezirkIDUndMeldungsart as Id in Repository
- use of Optional in Service and Repository
- some test improvements

dragonfly28 committed:
use @Query in repo

dragonfly28 committed:
Null-Check for wahlUndBezirkIDUndMeldungsart added

dragonfly28 committed:
List mapping replaced with stream.map.toList

dragonfly28 committed:
DTO for enum Meldungsart added

dragonfly28 committed:
timestamp 'erstelltAm' is set in service now and no more in controller

dragonfly28 committed:
method ordering changed

dragonfly28 committed:
test to overwrite entity added

dragonfly28 committed:
positive test for valid id added

dragonfly28 committed:
use 'val' and 'final' consistently

dragonfly28 committed:
sec test enhanced

dragonfly28 committed:
make sure that MockedStatic is always closed

dragonfly28 committed:
Merge branch 'dev' into 193-lesen-und-schreiben-von-ausdrucken

dragonfly28 committed:
merge syntax error corrected

dragonfly28 committed:
versioning sequence corrected

dragonfly28 committed:
test corrected

dragonfly28 committed:
mvn spotless

dragonfly28 committed:
Merge branch 'dev' into 193-lesen-und-schreiben-von-ausdrucken

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckModel.java:
```diff
@@ -0,0 +1,11 @@
+
+public record AusdruckModel(@Valid @NotNull WahlUndBezirkIDUndMeldungsart wahlUndBezirkIDUndMeldungsart,
+                            String content,
+                            Instant erstelltAm) {
```
Wie besprochen eine follow-Up-Issue machen damit `erstelltAm` nicht mehr Teil des Models ist

dragonfly28 replied:
Issue erstellt -> https://github.com/it-at-m/Wahllokalsystem/issues/759

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java:
```diff
@@ -0,0 +1,295 @@
+        }
+
+        @Test
+        void should_returnBadRequest_when_requestIsInvalid() throws Exception {
+
+            val request = MockMvcRequestBuilders.get(buildGetPostAusdruckURI(" ", " ", Meldungsart.V1));
```
```suggestion
        void should_returnBadRequest_when_requestIsInvalid() throws Exception {
            val request = MockMvcRequestBuilders.get(buildGetPostAusdruckURI(" ", " ", Meldungsart.V1));
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java:
```diff
@@ -0,0 +1,151 @@
+        }
+
+        @Test
+        void should_returnNullBodyWithHttpStatusNotFound_when_serviceReturnsEmtyOptional() {
```
```suggestion
        void should_returnNullBodyWithHttpStatusNotFound_when_serviceReturnsEmptyOptional() {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckServiceTest.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Nested
+    class GetAllAusdrucke {
```
bitte einen Test ergänzen der zeigt dass wenn keine Daten aus dem Repo kommen, dass eine leere Liste zurückkommt

dragonfly28 replied:
resolved with a66e91383e88fb577bb0641aead16466d25a7049

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckServiceTest.java:
```diff
@@ -0,0 +1,246 @@
+        }
+
+        @Test
+        void should_throwFachlicheWlsExceptionForParameter_when_callingValidator() {
```
```suggestion
        void should_throwFachlicheWlsExceptionForParameter_when_validatorValidationReturnsSetWithViolations() {
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/WahlUndBezirkIDUndMeldungsartValidatorTest.java:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        @Test
+        void should_throwFachlicheWlsException_when_wahlUndBezirkIDUndMeldungsartIsNull() {
```
kannst du den Fall nicht in den Argumentsstream einfließen lassen?

dragonfly28 replied:
resolved with b16320d7954e7f1cdc1d71d6c1522bfdd5567bd0

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerIntegrationTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/ausdruck/AusdruckControllerTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/ausdruck/AusdruckServiceTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
test for empty list added

dragonfly28 committed:
null argument added to stream

dragonfly28 committed:
Merge branch 'dev' into 193-lesen-und-schreiben-von-ausdrucken
